### PR TITLE
Merge merge_tryout into stable cards

### PR DESCRIPTION
**Pull request by JohanWulff**
- 'nuisance-aware' flat-s binning
- improved plots of final categories

### CODE REVIEW

File: tautaunn/tasks/datacards.py [hunk 1]
```diff
@@ -504,25 +632,26 @@ def card_names(self):
...
+
     def output(self):
         # prepare the output directory
         dirname = f"{self.binning}{self.n_bins}"
-        if self.binning in ["ud", "ud_flats", "tt_dy_driven"]:
-            dirname += f"_{self.uncertainty}_{self.signal_uncertainty}"
         if self.output_suffix not in ("", law.NO_STR):
             dirname += f"_{self.output_suffix.lstrip('_')}"
         d = self.local_target(dirname, dir=True)
+        # hotfix location in case TN_STORE_DIR is set to Marcel's
```
Comment (riga) [inline tautaunn/tasks/datacards.py]: @JohanWulff We removed this hotfix for now but feel free to change the "DEFAULT_STORE" attribute of the task to your value.

File: tautaunn/tasks/datacards.py [hunk 2]
```diff
@@ -541,21 +670,34 @@ def run(self):
 
         # prepare skim and eval directories, and samples to use per
         skim_directories = defaultdict(list)
+        # hardcode the eval directories for now
```
Comment (riga) [inline tautaunn/tasks/datacards.py]: We also used the computed requirements here.

File: tautaunn/write_datacards.py
Comment (riga) [inline tautaunn/write_datacards.py:1]: Do we still need this file?

File: tautaunn/write_datacards_stack.py [hunk 3]
```diff
@@ -2404,7 +1454,7 @@ def write(path):
     os.chmod(abs_datacard_path, 0o664)
 
     # return output paths
-    return abs_datacard_path, abs_shapes_path, bin_edges
+    return abs_datacard_path, abs_shapes_path, bin_edges, stop_reason#, bin_counts
```
Comment (riga) [inline tautaunn/write_datacards_stack.py]: fyi: we removed the comment, feel free to resolve :)

File: tautaunn/write_datacards_stack.py [hunk 4]
```diff
@@ -1812,184 +964,86 @@ def sort_values_and_weights(values, weights, inc=True):
...
+            
+            if len(hh_values) == 0:
+                print(f"no signal events found in ({category},{spin},{mass})")
+                bin_edges, stop_reason, bin_counts = [0., 1.], "no signal events found", None
+            else:
+                bin_edges, stop_reason, bin_counts = flats_systs(hh_shifts=hh_shifts,
+                                                                 tt_shifts=tt_shifts,
+                                                                 dy_shifts=dy_shifts,
+                                                                 all_bkgds=(all_bkgds_values, all_bkgds_weights),
+                                                                 error_target=1, #0.25, # changed from 1
```
Comment (riga) [inline tautaunn/write_datacards_stack.py]: We stumbled over this error_target and tried to see where it is used ... is it used anywhere ? 😅

File: tautaunn/binning_algorithms.py [hunk 5]
```diff
@@ -1,54 +1,83 @@
...
+
+
+def check_error_requirement(bkgd_values, bkgd_weights, next_edge, target_val=1.):
+    mask = bkgd_values >= next_edge
+    return calc_rel_error(bkgd_weights[mask]) < target_val
+
+
+def get_conditions(dy_vals_weights, tt_vals_weights, next_edge, yield_target=1e-5, error_target=1.):
+    dy_conds = (check_yield_requirement(*dy_vals_weights, next_edge)
+                and check_error_requirement(*dy_vals_weights, next_edge))
```
Comment (riga) [inline tautaunn/binning_algorithms.py]: Shouldn't `{yield,error}_target` be forwarded to `check_{yield,error}_requirement`?

File: tautaunn/write_datacards_stack.py (hunk 4 above)
Comment (riga) [reply tautaunn/write_datacards_stack.py]: We commented below where we think this should be introduced.

**Review by riga — CHANGES_REQUESTED** [review]

File: tautaunn/write_datacards.py
Comment (JohanWulff) [reply tautaunn/write_datacards.py:1]: no

File: tautaunn/binning_algorithms.py (hunk 5 above)
Comment (JohanWulff) [reply tautaunn/binning_algorithms.py]: done

File: tautaunn/write_datacards_stack.py (hunk 4 above)
Comment (JohanWulff) [reply tautaunn/write_datacards_stack.py]: yeah this was indeed not forwarded so far... it should be done now

**Comment by JohanWulff** [comment]
There's a few files (everything that was done for the alternative way to write the datacards (jows_cards.py)) which don't really have to be merged into main. Shall i remove these from the PR first?

**Comment by riga** [comment]
I guess there's no harm in keeping them. In case we clean the code base towards the end, we can eventually drop them.

**Pull request merged**